poppy/poppy_core.py

```diff
@@ -687,21 +687,42 @@ def _resample_wavefront_pixelscale(self, detector):
         _log.debug("Desired detector FOV: {} pixels, {:.3f}".format(detector.shape,
                                                                     detector.shape[0]*u.pixel*detector.pixelscale))
 
-        # TODO the following works but is not optimal for performance
-        # We should consider cropping out an appropriate subregion prior to performing the zoom.
-        # That makes a difference if the detector is only sampling a small part of a much larger wavefront
-
-        new_wf_real = scipy.ndimage.zoom(self.wavefront.real, pixscale_ratio, order=detector.interp_order)
-        new_wf_imag = scipy.ndimage.zoom(self.wavefront.imag, pixscale_ratio, order=detector.interp_order)
-        new_wf = new_wf_real + 1.j*new_wf_imag
+        def make_axis(npix, step):
+            """ Helper function to make coordinate axis for interpolation """
+            return step * np.arange(-npix // 2, npix // 2, dtype=np.float64)
+
+        # Provide 2-pixel margin around image to reduce interpolation errors at edge, but also make
+        # sure that image is centered properly after it gets cropped down to detector size
+        margin = 2
+        crop_shape = [margin + shape for shape in self.wavefront.shape]
+
+        # Crop wavefront down to detector size + margin- don't waste computation interpolating
+        # parts of plane that get cropped out later anyways
+        cropped_wf = utils.pad_or_crop_to_shape(self.wavefront, crop_shape)
+
+        # Input and output axes for interpolation.  The interpolated wavefront will be evaluated
+        # directly onto the detector axis, so don't need to crop afterwards.
+        x_in = make_axis(crop_shape[0], self.pixelscale)
+        y_in = make_axis(crop_shape[1], self.pixelscale)
+        x_out = make_axis(detector.shape[0], detector.pixelscale)
+        y_out = make_axis(detector.shape[1], detector.pixelscale)
```

Comment:
Hi @sdwill, in my testing I found one issue here. The pixel scales need be converted to the same units here - this covers the case where e.g. the detector pixel scale is specified in microns/pixel but the wavefront is in meters/pixel. 

```
        x_in = make_axis(crop_shape[0], self.pixelscale.to(u.m/u.pix).value)
        y_in = make_axis(crop_shape[1], self.pixelscale.to(u.m/u.pix).value)
        x_out = make_axis(detector.shape[0], detector.pixelscale.to(u.m/u.pix).value)
        y_out = make_axis(detector.shape[1], detector.pixelscale.to(u.m/u.pix).value)
```

Reply:
That's fair, and I admit I overlooked it during implementation.